docs/modules/ROOT/pages/writing-upgradeable.adoc

```diff
@@ -0,0 +1,394 @@
+
+[source,solidity]
+----
+// @openzeppelin/contracts/token/ERC20/ERC20.sol
+pragma solidity ^0.6.0;
+
+  ...
+
+contract ERC20 is Context, IERC20 {
+
+  ...
+
+    string private _name;
+    string private _symbol;
+    uint8 private _decimals;
+
+    constructor (string memory name, string memory symbol) public {
+        _name = name;
+        _symbol = symbol;
+        _decimals = 18;
+    }
+
+  ...
+}
+----
+
+This means you should not be using these contracts in your OpenZeppelin Upgrades project. Instead, make sure to use `@openzeppelin/contracts-ethereum-package`, which is an official fork of OpenZeppelin Contracts that has been modified to use initializers instead of constructors. Take a look at how https://github.com/OpenZeppelin/openzeppelin-contracts-ethereum-package/blob/v3.0.0/contracts/token/ERC20/ERC20.sol[ERC20] looks in `@openzeppelin/contracts-ethereum-package`:
+
+[source,solidity]
+----
+// @openzeppelin/contracts-ethereum-package/contracts/token/ERC20/ERC20.sol
+pragma solidity ^0.6.0;
+  ...
+contract ERC20UpgradeSafe is Initializable, ContextUpgradeSafe, IERC20 {
+  ...
+    string private _name;
+    string private _symbol;
+    uint8 private _decimals;
+
+    function __ERC20_init(string memory name, string memory symbol) internal initializer {
+        __Context_init_unchained();
+        __ERC20_init_unchained(name, symbol);
+    }
+
+    function __ERC20_init_unchained(string memory name, string memory symbol) internal initializer {
+        _name = name;
+        _symbol = symbol;
+        _decimals = 18;
+    }
+  ...
+}
+----
+
+Whether using OpenZeppelin Contracts or another Ethereum Package, always make sure that the package is set up to handle upgradeable contracts.
+
+[[avoid-initial-values-in-field-declarations]]
+=== Avoiding Initial Values in Field Declarations
+
+Solidity allows defining initial values for fields when declaring them in a contract.
+
+[source,solidity]
+----
+contract MyContract {
+    uint256 public constant hasInitialValue = 42;
+}
+----
+
+This is equivalent to setting these values in the constructor, and as such, will not work for upgradeable contracts. Make sure that all initial values are set in an initializer function as shown below; otherwise, any upgradeable instances will not have these fields set.
+
+[source,solidity]
+----
+contract MyContract is Initializable {
+    uint256 public hasInitialValue;
+
+    function initialize() public initializer {
+        hasInitialValue = 42;
+    }
+}
+----
+
+Note that it still is fine to set constants here, because the compiler https://solidity.readthedocs.io/en/latest/contracts.html#constant-state-variables[does not reserve a storage slot for these variables], and every occurrence is replaced by the respective constant expression. So the following still works with OpenZeppelin Upgrades:
```

Comment:
```suggestion
Note that it is still fine to define _constant_ state variables in this way, because the compiler https://solidity.readthedocs.io/en/latest/contracts.html#constant-state-variables[does not reserve a storage slot for these variables], and every occurrence is replaced by the respective constant expression. So the following still works with OpenZeppelin Upgrades:
```